### PR TITLE
Reaction graph

### PR DESCRIPTION
**Pull request by edwardalee**
This branch merges the various implementations of graphs representing dependencies between reactions into one so that the complexity of handling banks and multiports can be handled in one unified place.

### CODE REVIEW

File: org.lflang/src/org/lflang/generator/MultiportInstance.xtend [hunk 1]
```diff
@@ -86,6 +88,24 @@ class MultiportInstance extends PortInstance {
         }
     }
     
+    /////////////////////////////////////////////
+    //// Public Fields
+
+    /** The array of instances. */ 
+    public val instances = new LinkedHashSet<PortInstance>()
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/MultiportInstance.xtend:95]: Should this be `public`?

**Review by lhstrh — APPROVED** [review]
Looks good! I'm exited that we got to address this so quickly!

File: org.lflang/src/org/lflang/DefaultErrorReporter.xtend [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package org.lflang
+
+import org.eclipse.emf.ecore.EObject
+
```
[the remark this answers is not in the document]
Comment (lhstrh) [reply org.lflang/src/org/lflang/DefaultErrorReporter.xtend:4]: See 7d0d1e8c.

**Comment by lhstrh** [comment]
I've noticed that the deprecated methods in `ASTUtils` are still present and still used in the validator. We should be able to eliminate them...

**Comment by edwardalee** [comment]
Yes, I just realized that too...

I guess we should not touch the CppGenerator?

Edward

---------
Edward A. Lee
EECS, UC Berkeley
***@***.***
http://eecs.berkeley.edu/~eal

**Comment by cmnrd** [comment]
Are you referring to the deprecated methods in ASTUtils that deal with port widths?  I think it is safe to remove them. The C++ generator in Kotlin uses the def `static int width(WidthSpec spec, List<Instantiation> instantiations)` method (but passes `null` to the instantiations argument).

**Comment by edwardalee** [comment]
I just realized we aren't anywhere near done yet. The deprecated methods are still being used and the Generator still duplicates code from ReactorInstance.  Working on it.

**Comment by lhstrh** [comment]
Yes, that's what I meant with my earlier comment... Better to fix that before merging...

**Comment by edwardalee** [comment]
Deprecated methods have been removed from ASTUtils. The CppGenerator uses the non-deprecated methods with incomplete arguments, and GeneratorBase still (incompletely) duplicates some of the bank/multiport logic in analyzeFederates, so there is still work to do, but this already fixes some bugs and all tests seem to be passing.

**Pull request merged**